bionic/task_state.py

```diff
@@ -0,0 +1,306 @@
+from .cache import Provenance
+from .datatypes import ProvenanceDigest, Query, Result
+from .exception import CodeVersioningError
+from .util import oneline
+
+
+class TaskState:
+    """
+    Represents the state of a task computation.  Keeps track of its position in
+    the task graph, whether its values have been computed yet, additional
+    intermediate state and the deriving logic.
+    """
+
+    def __init__(self, task, dep_states, case_key, provider):
+        self.task = task
+        self.dep_states = dep_states
+        self.case_key = case_key
+        self.provider = provider
+
+        # These are set by complete(), just
+        # before the task state becomes eligible for cache lookup / computation.
+        #
+        # They will be present if and only if is_complete is True.
+        self.is_initialized = False
```

Comment:
`is_initialized` will be present even if `is_complete` is false, right?

Reply:
Oh yeah, good catch! This comment is now outdated. I'll update it.